tools/elfdeps.c

```diff
@@ -33,6 +38,97 @@ typedef struct elfInfo_s {
     ARGV_t provides;
 } elfInfo;
 
+/*
+ * If filename contains ".so" followed by a version number, return
+ * a copy of the version number.
+ */
+static char *getLibtoolVer(const char *filename)
+{
+    const char *so;
+    char dest[PATH_MAX];
+    int destsize = 0;
+    int found_digit = 0, found_dot = 0;
+
+    destsize = readlink(filename, dest, PATH_MAX);
+    if (destsize > 0) {
+	dest[destsize] = 0;
+	filename = dest;
+    }
+    // Start from the end of the string.  Verify that it ends with
+    // numbers and dots, preceded by ".so.".
+    so = filename + strlen(filename);
+    while (so > filename+2) {
+	if (*so == '.') {
+	    found_dot++;
+	    so--;
+	    continue;
+	} else if (strchr("0123456789", *so)) {
+	    found_digit++;
+	    so--;
+	    continue;
+	} else if (strncmp(so-2, ".so.", 4) == 0) {
+	    so+=2;
+	    if (found_digit && found_dot > 1) {
+		return strdup(so);
+	    }
+	    break;
+	} else {
+	    break;
+	}
+    }
+    return NULL;
+}
+
+/*
+ * Rather than re-implement path searching for shared objects, use
+ * dlmopen().  This will still perform initialization and finalization
+ * functions, which isn't necessarily safe, so do that in a separate
+ * process.
+ */
+static char *getLibtoolVerFromShLink(const char *filename)
+{
+    char dest[PATH_MAX];
+    int pipefd[2];
+    pid_t cpid;
+
+    if (pipe(pipefd) == -1) {
+	return NULL;  // Should this be a fatal error instead?
+    }
+    cpid = fork();
+    if (cpid == -1) {
+	return NULL;  // Should this be a fatal error instead?
+    }
+    if (cpid == 0) {
+	void *dl_handle;
+	struct link_map *linkmap;
+	char *version = NULL;
+
+	close(pipefd[0]);
+	dl_handle = dlmopen(LM_ID_NEWLM, filename, RTLD_LAZY);
+	if (dl_handle == NULL) _exit(0);
+	if (dlinfo(dl_handle, RTLD_DI_LINKMAP, &linkmap) != -1) {
```

Comment:
Sadly both dlinfo() and dlmopen() are GNU extensions, so these need to be checked for in cmake and this works made conditional on their availability.